### PR TITLE
Checkout: Fix cart icon focus state

### PR DESCRIPTION
**Pull request by vindl**
Resolves: https://github.com/Automattic/wp-calypso/issues/21659

Adds focus styles for keyboard-only access to `cart` icon and individual `remove-item` icons.

### Visual changes

**Before**
- There were no style changes when cart was focused, and tabbing would skip the cart contents.

before

**After**
- Focus styles for cart icon and contents are visible and the focus is kept in popover component while it's open.

after

### Testing instructions
- You can follow the instructions available in https://github.com/Automattic/wp-calypso/issues/21659 to replicate.

### CODE REVIEW

File: client/components/popover/index.jsx [hunk 1]
```diff
@@ -118,12 +120,17 @@ class Popover extends Component {
...
 
 	componentDidUpdate( prevProps, prevState ) {
-		const { isVisible } = this.props;
+		const { isVisible, onShow, focusOnShow } = this.props;
 
 		if ( ! prevState.show && this.state.show ) {
 			this.bindEscKeyListener();
 			this.bindDebouncedReposition();
 			bindWindowListeners();
+			onShow();
```
Comment (vindl) [inline client/components/popover/index.jsx]: Note: this is unrelated to the current issue, but the `onShow` prop is documented in components readme, but it was not used/called anywhere.

**Comment by vindl** [comment]
> Try to tab to the checkout button. You can't - not without tabbing through the entire page and sidebar, which is not expected or intuitive.

I'm having trouble with the last part of the issue. I tried forcing the focus on Popover container manually and adding `tabIndex="0"` to individual `remove-item` icons, but the focus still switches to the elements in the background, and only comes back to cart buttons in the end.

Any ideas on how to resolve this? Also I think this is not specific to `PopoverCart` component and happens with `Popover` in general.

File: client/components/popover/index.jsx (hunk 1 above)
Comment (gwwar) [reply client/components/popover/index.jsx]: I assume this is an optional handler passed by the parent component.

Comment (vindl) [reply client/components/popover/index.jsx]: @gwwar yes, but passing the handler would do nothing since it was never executed.

Comment (gwwar) [reply client/components/popover/index.jsx]: Ah I see what you're saying now. Let's leave this for a different janitorial. We can do an audit on usages, and if no one actually is using this, we can remove the prop and update the readme.

Comment (vindl) [reply client/components/popover/index.jsx]: Sure, removed.

**Comment by kwight** [comment]
Works great, as advertised 👍

I'm a little hesitant to add a new, little-known module, but fixing this "properly" would mean moving the popup markup within our page structure which would certainly need much investigation and testing. Anyone else feel strongly one way or another?

**Comment by kwight** [comment]
(Oh, I was also getting errors when trying to test, but a rebase and fresh start solved it.)

**Comment by gwwar** [comment]
@jsnajdr any thoughts on the bump to the build bundle? The module can potentially be used in other areas for example enabling focus trapping in our dialog modals when they're visible.

It does feel like the popover/cart can in the future be async loaded.

**Comment by retrofox** [comment]
It works like a charm. I think it's a good solution for a tricky issue.

> This should fix the same issue for Popover components in general and it might be worthwhile to pull it out in a separate PR.

The Popover container element is injected into the root (child of the body) which means it won't follow the same focusing order that the user hopes according to the visual order shown by the interface.

LGTM.

**Comment by ryelle** [comment]
This is great for the keyboard accessibility part of things! Thanks for taking it on 😄 

I did notice that it's not possible to close the popover unless you check out - even emptying the cart leaves you with a cart popover:

screen shot 2018-02-22 at 4 53 39 pm

This also doesn't handle the VoiceOver part of things well - first, you initially land focus onto the X icon, which doesn't have a label so it just reads "button". It's also possible to navigate out of the popover with VoiceOver, and then it's very hard to get back in, but the focus stays in the popover, so it's just broken. Would it be possible to use `react-modal` (what we use for the Dialog component) here? That's already solved the focus trap and screen reader issues.

**Comment by vindl** [comment]
> I did notice that it's not possible to close the popover unless you check out - even emptying the cart leaves you with a cart popover.

It's possible to close it with `Esc` button - is that insufficient in this case?

> Would it be possible to use react-modal (what we use for the Dialog component) here? That's already solved the focus trap and screen reader issues.

Not sure yet, I'll have to investigate that option.

**Comment by jsnajdr** [comment]
> any thoughts on the bump to the build bundle?

@gwwar That's the code of the new dependency (`focus-trap-react`) that's used in the general purpose `Popover` component.

> It does feel like the popover/cart can in the future be async loaded.

Yes, the shopping cart code is certainly a great candidate to be async loaded. Currently, during initial load of any route, the cart code is initialized and the cart contents are requested from an endpoint. So that it can bug you with notices and popups about an unfinished purchase, wherever in Calypso you are.

At least, the cart initialization could be postponed a few seconds, so that it doesn't block the initial render of the content the user is actually interested in. Ad trackers commonly behave this way (we load a few of them in Calypso to measure performance of WP.com ads)

**Comment by ryelle** [comment]
> It's possible to close it with Esc button - is that insufficient in this case?

Right, but if someone is using a mouse, it's surprising that they can't click outside of the popover to close it. A mouse-user might not jump to trying the keyboard.

**Comment by gwwar** [comment]
> @gwwar That's the code of the new dependency (focus-trap-react) that's used in the general purpose Popover component.

@jsnajdr Oh I meant, thoughts on how we feel if it is worth pulling in for the size/use cases :D

Though it sounds like we have a few more cases to work through here.

**Comment by jsnajdr** [comment]
> Oh I meant, thoughts on how we feel if it is worth pulling in for the size/use cases

I'd say that the new dependencies do useful work and they are not bloated in any way. `focus-trap-react`, `focus-trap` and `tabbable` are reasoable modules.

**Comment by vindl** [comment]
> Right, but if someone is using a mouse, it's surprising that they can't click outside of the popover to close it. A mouse-user might not jump to trying the keyboard.

Oh right, I was so focused on keyboard-only testing that I completely forgot to check everything again with mouse only. :)

**Comment by vindl** [comment]
> Would it be possible to use react-modal (what we use for the Dialog component) here? That's already solved the focus trap and screen reader issues.

@ryelle I looked into this but it seems like it would require a more involved refactor and it wouldn't really match the popover in terms of appearance and positioning. 

I removed the changes related to `react-focus-trap` and just left the styling updates. This solves part of the issue at least, and I liked to get it merged if possible. We might circle back to Popover component fix but before that the plan is to prioritize other bugs in our backlog.

**Comment by kwight** [comment]
> This solves part of the issue at least, and I liked to get it merged if possible.

The styling updates do take care of the first stage of the problem (going from Plans -> Domains -> Email -> Cart). Unfortunate the next isn't Checkout after Enter (focus will proceed to plan Upgrade buttons), but a solid solution (whether it's markup re-arrangement, `react-focus-trap`, `tabindex` wars, or a combination) could use more examination in another PR. 👍

**Comment by vindl** [comment]
> Unfortunate the next isn't Checkout after Enter (focus will proceed to plan Upgrade buttons), but a solid solution (whether it's markup re-arrangement, react-focus-trap, tabindex wars, or a combination) could use more examination in another PR.

I created a separate issue for this here https://github.com/Automattic/wp-calypso/issues/23120.

**Pull request merged**